addons/udes_security/tests/test_allowed_file_type.py

```diff
@@ -0,0 +1,148 @@
+                "datas": base64.b64encode(file_path.read_bytes()),
+            }
+        )
+        return attachment
+
+    @classmethod
+    def update_attachment(self, attachment, filename):
+        """Update attachment data from supplied filename"""
+        file_path = self._get_file_path(filename)
+
+        res = attachment.write({
+            "name": file_path.name,
+            "datas_fname": file_path.name,
+            "datas": base64.b64encode(file_path.read_bytes()),
+        })
+
+        return res
+
+    @classmethod
+    def create_allowed_file_type(self, file_type, **kwargs):
+        """Create allowed file type record from supplied file type"""
+        allowed_file_type_vals = {"name": file_type}
+        allowed_file_type_vals.update(kwargs)
+
+        allowed_file_type = self.AllowedFileType.create(allowed_file_type_vals)
+
+        return allowed_file_type
+
+    def test_allowed_file_active_status(self):
+        """Assert that an attachment is active if its file type is allowed, otherwise inactive"""
+        # Check csv file is active and txt file is inactive
+        self.assertTrue(self.atch_csv_file.active)
+        self.assertFalse(self.atch_txt_file.active)
+
+        # Set csv allowed file type to inactive and assert csv attachment is now inactive
+        self.allowed_file_type_csv.active = False
+        self.assertFalse(self.atch_txt_file.active)
+
+        # Check that txt file is unaffected and still inactive
+        self.assertFalse(self.atch_txt_file.active)
+
+    def test_allowed_file_type_deleted(self):
+        """
+        Assert that deleting an active allowed file type sets attachments of that file type
+        to inactive
+        """
+        # Assert that csv file is inactive after csv allowed file type record is deleted
+        self.allowed_file_type_csv.unlink()
+        self.assertFalse(self.atch_csv_file.active)
+
+    def test_assert_allowed_file_type_unique(self):
+        """Assert that only one blocked file type record can exist for each file type"""
+        # Create another allowed file type for csv and assert that an Integrity Error is raised
+        with mute_logger("odoo.sql_db"), self.assertRaises(IntegrityError):
+            self.create_allowed_file_type(self.allowed_file_type_csv.name)
+
+    def test_assert_allowed_file_type_name_change_prevented(self):
+        """Assert that the file type (aka name) on an allowed file type record cannot be changed"""
+        with self.assertRaises(UserError):
+            self.allowed_file_type_csv.name = "other_file_type"
+
+    def test_assert_allowed_file_type_names_formatted(self):
+        """
+        Assert that file type name is properly formatted, 
+        i.e. lowercase with full stop character removed
+        """
+        json_file_type = "json"
+        json_unformatted_file_types = ("JSON", ".json", ".JSON")
+
+        # Ensure that _format_name properly formats each unformatted file type
```

Comment:
Consider using a [subtest](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) here.

Reply:
Thanks for the tip! Am using subtest now and looks much better. Also much nicer to debug if it fails.

Will have to keep that in mind for future. :)

Reply:
Most of my code reviews of test code are "You could use subtest here" - for some reason no-one seems to know about it :shrug: